[SYCL][ESIMD] Add test on simd converting constructor

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add test on simd converting constructor

vasilytric committed:
Merge branch 'intel' into simd_ctor_conv

vasilytric committed:
[SYCL][ESIMD] Update code to use new functions

- Change run_test signature
- Add add expected test cases
- Update comments

vasilytric committed:
[SYCL][ESIMD] Update verification logic for converting from fp to fp

This update required for case when current value is demoralized because
the rule of rounding on the host and device sides is undefined.

vasilytric committed:
[SYCL][ESIMD] Change type of values to destination while logging

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_converting.hpp:
```diff
@@ -0,0 +1,164 @@
+
+namespace esimd_test::api::functional::ctors {
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct initializer {
```
I see `initializer` context only. Don't we need to cover other contexts?

vasilytric replied:
> Don't we need to cover other contexts?

I don't know how to cover other contexts. could you help me with that?

vasilytric replied:
the other contexts has been added [here](https://github.com/intel/llvm-test-suite/pull/695/commits/7b72a178bbd699b5717065a05b91c8a76bad0816#diff-e782ba6c3846832a7b32bf3abe80d4517ba0bc9c99a52c6baffe5f19fedcb46eR38-R92)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_converting.hpp:
```diff
@@ -0,0 +1,164 @@
+    queue.wait_and_throw();
+
+    for (size_t i = 0; i < result.size(); ++i) {
+      const DstT &expected = static_cast<DstT>(ref_data[i]);
```
```suggestion
      // We ensure there is no UB here by preparing appropriate reference values
      const DstT &expected = static_cast<DstT>(ref_data[i]);
```

vasilytric replied:
This comment has been applied in [here](https://github.com/intel/llvm-test-suite/pull/695/commits/7b72a178bbd699b5717065a05b91c8a76bad0816#diff-e782ba6c3846832a7b32bf3abe80d4517ba0bc9c99a52c6baffe5f19fedcb46eR180-R181)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_converting.hpp:
```diff
@@ -0,0 +1,164 @@
+        // https://github.com/microsoft/STL/issues/519
+        if (!std::isnan(expected) && !std::isnan(retrieved)) {
+          if (expected != retrieved) {
+            if (!((expected + value<DstT>::pos_ulp(expected)) >= retrieved ||
```
expected `100`, retrieved `nan` - gives false positive
expected `nan`, retrieved `100` - gives false positive

I would suggest to fix the L132 by either:
```
if (!(std::isnan(expected) && std::isnan(retrieved))) {
```
or
```
if (!std::isnan(expected) || !std::isnan(retrieved)) {
```

Or maybe it's better from the point of readability to have something like this:
```
if (std::isnan(expected) && std::isnan(retrieved) {
  // Having NaN with any opcode inside is OK
} else if (expected == retrieved) {
  // Values are equal, no more checks required
} else if (is_within_range(retrieved, ulp_range(expected, 1))) {
  // Comment with the reasons behind such check
} else {
  passed = fail_test(...);
}
```

vasilytric replied:
> if (!std::isnan(expected) || !std::isnan(retrieved)) {

I have applied this comment [here](https://github.com/intel/llvm-test-suite/pull/695/commits/7b72a178bbd699b5717065a05b91c8a76bad0816#diff-e782ba6c3846832a7b32bf3abe80d4517ba0bc9c99a52c6baffe5f19fedcb46eR188)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_converting.hpp:
```diff
@@ -0,0 +1,164 @@
+        if (!std::isnan(expected) && !std::isnan(retrieved)) {
+          if (expected != retrieved) {
+            if (!((expected + value<DstT>::pos_ulp(expected)) >= retrieved ||
+                  (expected - value<DstT>::pos_ulp(expected)) >= retrieved ||
+                  (expected + value<DstT>::pos_ulp(expected)) <= retrieved ||
```
Isn't there a possibility to have:

- either `expected == value<DstT>::max()`
- or `expected == value<DstT>::lowest()`

here in future? Any of these would result in UB. Please add `TODO` note with assertions at least.

Also please note that `expected - value<DstT>::pos_ulp(expected)` is not equal to `expected - value<DstT>::neg_ulp(expected)`, so you need to fix the conditions.
I would suggest to use `std::nextafter`/`std::nexttoward` directly, but it requires a new method to `value<T>` to cover the `half` type.

vasilytric replied:
TODO comment has been added [here](https://github.com/intel/llvm-test-suite/pull/695/commits/7b72a178bbd699b5717065a05b91c8a76bad0816#diff-e782ba6c3846832a7b32bf3abe80d4517ba0bc9c99a52c6baffe5f19fedcb46eR190-R193)

yuriykoch replied:
Thanks :)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_converting.hpp:
```diff
@@ -0,0 +1,164 @@
+    // one value from reference data.
+    // If current number of elements is greater than one, then run tests with
+    // whole reference data.
+    if constexpr (NumElems == 1) {
```
Minor: have you verified compilation/run for this condition? Looks like it's always false with current test coverage. It's great to have a possibility to switch support for `simd<T,1>` by single line in tests though.

What would you say about using the `ESIMD_TESTS_FULL_COVERAGE` in cpp files? So we could verify test works for `simd<T,1>` at any time by simple macro definition.

vasilytric replied:
I have verified this and we don't have any syntaxes error in this branch, also I don't wont to add a macros to enable this branch and I don't won't to delete this branch, it may be useful in the future

vasilytric committed:
[SYCL][ESIMD] Small updates

- Update comments, add TODO comment
- Verify that one value is nan, instead of verify tha both values is
  nan
- Add additional constructor calling contexts

vasilytric committed:
Merge branch 'intel' into simd_ctor_conv

vasilytric commented:
@v-klochkov Would you mind to take a look?

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_converting_core.cpp:
```diff
@@ -0,0 +1,66 @@
+  const auto uint_types = get_tested_types<tested_types::uint>();
+  const auto sint_types = get_tested_types<tested_types::sint>();
+  const auto core_types = get_tested_types<tested_types::core>();
+  const auto single_size = get_sizes<8>();
```
Would you please verify '1' as well. Scalar is often handled differently.

vasilytric replied:
The verification for `1` has been added in this commit: https://github.com/intel/llvm-test-suite/pull/695/commits/22de452fa8e6f6cb13eccf7e7715767328638e78

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_converting_fp_extra.cpp:
```diff
@@ -0,0 +1,71 @@
+// The current "REQUIRES" should be replaced with "gpu" only as mentioned in
+// "XREQUIRES".
+// UNSUPPORTED: cuda, hip
+// XRUN: %clangxx -fsycl %s -fsycl-device-code-split=per_kernel -o %t.out
```
The test is not compilable, I guess. Would you please add either an internal tracker or an issue?

vasilytric replied:
This test has been enabled in this commit: https://github.com/intel/llvm-test-suite/pull/695/commits/900b37e35bcdebdd0a2ca84f486fb267455631c7

vasilytric committed:
[SYCL][ESIMD] Add coverage for 1 simd size

vasilytric committed:
[SYCL][ESIMD] Enable test for fp_extra types

vasilytric committed:
Merge branch 'intel' into simd_ctor_conv

vasilytric committed:
[SYCL][ESIMD] Use new esimd namespace
